dropwizard-hibernate/src/main/java/io/dropwizard/hibernate/UnitOfWorkAspect.java

```diff
@@ -68,19 +71,24 @@ public void beforeStart(@Nullable UnitOfWork unitOfWork) {
             existingSession = sessionFactory.getCurrentSession();
         }
 
-        if (existingSession != null) {
+        if (existingSession != null && existingSession.isOpen() && existingSession.getSessionFactory() != null) {
+            logger.info("Reusing existing session");
             sessionCreated = false;
         } else {
-            Session session = sessionFactory.openSession();
-            sessionCreated = true;
+            Session session = null;
             try {
+                session = sessionFactory.openSession();
```

Comment:
Moved more of this into the try/catch. Also move the context cleanup to a finally to make sure it happens